### PR TITLE
[vcl] don't set TTL to 0

### PR DESCRIPTION
**Pull request by gquintard**
Because of request coalescing, setting the TTL to 0s actually causes
backend request queueing, meaning that for one specific object,
Varnish will only allow one request at a time.

For Varnish 4, set the TTL to 2 minutes, leveraging Hit-for-Pass and
Varnish will transform any hit into a pass for the next 2 minutes,
preventing queueing.

For 5 and 6, the default mechanism is Hit-for-Miss, which works the same
as HfP with the added bonus that if the response changes and we we decide
to cache it, this will override the HfM TTL. So we can set the TTL for
one day.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by gquintard** [comment]
@magento run all tests

File: app/code/Magento/PageCache/etc/varnish4.vcl
Comment (lbajsarowicz) [inline app/code/Magento/PageCache/etc/varnish4.vcl:169]: What is the purpose of caching 500 / 503 / 403?

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]

File: app/code/Magento/PageCache/etc/varnish4.vcl
Comment (gquintard) [reply app/code/Magento/PageCache/etc/varnish4.vcl:169]: it doesn't cache them because of the following `set beresp.uncacheable = true;`. Instead Varnish will remember that these objects are uncacheable, and will disable request collapsing for them.

If you keep the TTL to 0, you get into something like this:
- 20 users request with the same object
- varnish only goes to the backend once (request coalescing)
- varnish realizes it cannot reuse the object, so it only serves it to one client
- we now have 19 users request with the same object
- varnish only goes to the backend once
- varnish realizes ....
rinse, repeat. Ultimately, the last request needs to wait 20 round-trip to the backend before being delivered.

If you set the TTL to some positive value AND the uncacheable bit to true:
- 20 users request with the same object
- varnish only goes to the backend once (request coalescing)
- varnish realizes it cannot reuse the object, so it only serves it to one client, but remembers it as uncacheable 
- the 19 remaining requests are woken up, and can all go to the backend in parallel

it's one of the trickiest Varnish gotchas, so please let me know if I wasn't clear.

Comment (lbajsarowicz) [reply app/code/Magento/PageCache/etc/varnish4.vcl:169]: @gquintard Your explanation sounds fine for me. :heavy_check_mark:

**Review by lbajsarowicz — APPROVED** [review]
:heavy_check_mark:  After discussing my concerns with the author.
Thank you for your expertise!

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests B2B

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7765](https://jira.corp.magento.com/browse/ENGCOM-7765) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
Hi @lbajsarowicz . Could you put appropriate labels for test coverage #28927 #28928 #28944 ?
Thanks!

**Comment by lbajsarowicz** [comment]
**Automated Tests** are not applicable, as the change is related to infrastructure configuration.

**Comment by engcom-Bravo** [comment]
Dev experience is required for testing of this PR. Please note that Manual testing has not been performed.

**Comment by engcom-Charlie** [comment]
HI, @gquintard can you please add some clarification on how can I check it by the manual scenario?
Thank you.

**Comment by gquintard** [comment]
@engcom-Charlie, I can provide a `varnishtest` file, but i assume you will want to to test with a full setup, so here's what you can do:
- create a page/endpoint that will take 10 seconds to respond with a 405 status (and a random string as body, just to be sure).
- send 10 `curl` on that endpoint

Before the patch, you should see each response coming in one at a time, 10 seconds after the previous one. So the last one will have a TTLB of 100 seconds.
After the patch, you should see everyone stalling for 10 seconds, get a first response back, then the 9 others 10 seconds after. (basically, varnish needs the first response to realize it shouldn't coalesce requests).

**Comment by gquintard** [comment]
Hum, actually, you are saved from request coalescing by the `grace` set above., so it sort of works.
I'd still prefer to get rid of the `ttl = 0s` as it's an example of cargo cult, and people are going to copy-paste that snippet without realizing that the `grace` 20 lines above is important, but as it is, it works.

**Comment by ihor-sviziev** [comment]
@engcom-Charlie could you check the reply from the PR author?

**Comment by gquintard** [comment]
@engcom-Charlie happy new year, and bump! :-)

**Review by ihor-sviziev — APPROVED** [review]

**Comment by ihor-sviziev** [comment]
Let's re-approve it

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7765](https://jira.corp.magento.com/browse/ENGCOM-7765) has been created to process this Pull Request

**Comment by gquintard** [comment]
hi, we are almost at the one-year anniversary of that PR, is there something I can do to make it happen?

**Comment by ihor-sviziev** [comment]
@gquintard, unfortunately, this PR has low priority and will be processed after all high priority ones

**Comment by hostep** [comment]
> this PR has low priority and will be processed after all high priority ones

This basically means never, because there will always be higher prio tickets.
I don't think this approach is good and the duration a PR has been open should be considered as well, regardless of priority, no?

**Comment by hostep** [comment]
See extra arguments in https://github.com/magento/magento2/pull/33604#issue-697778853 around this change. Maybe the priority should be increased here?

**Comment by bartoszkubicki** [comment]
Should be marked as top priority, it solve possible performance issue

**Comment by lbajsarowicz** [comment]
@gquintard I'm no longer a Magento Maintainer, but I'm really sorry you've been waiting for more than a year to get **any** feedback. As this PR became... popular (https://twitter.com/IvanChepurnyi/status/1420298850833682432), I'm pretty sure Magento Community Engineering has extra motivation to review and merge it.

For those who encounter the issue - there's always a way to generate Composer Patch: https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/28927.diff and apply that to your project directly without waiting years.

**Comment by ThijsFeryn** [comment]
If tests are required to prove the impact of this PR, here are some `varnishtest` files:

```
varnishtest "Testing beresp.ttl=0s vs beresp.ttl=120s"
server s1 {
	rxreq
	delay 5
	txresp -status 201 -body "1\n"
} -start
server s2 {
	rxreq
	delay 5
	txresp -status 201 -body "2\n"
} -start
server s3 {
	rxreq
	delay 5
	txresp -status 201 -body "3\n"
} -start
varnish v1 -vcl+backend {
	sub vcl_backend_fetch {
		if (bereq.http.be == "two") {
			set bereq.backend = s2;
		} else if (bereq.http.be == "three") {
			set bereq.backend = s3;
		}
	}
	sub vcl_backend_response {
		if (beresp.status != 200 && beresp.status != 404) {
			set beresp.ttl = 0s;
			set beresp.uncacheable = true;
			return (deliver);
		}
	}
} -start
client c1 {
	txreq
	rxresp
	expect resp.status == 201
	expect resp.body ~ "^1"
} -start
client c2 {
	txreq -hdr "be: two"
	rxresp
	expect resp.status == 201
	expect resp.body ~ "^2"
} -start
client c3 {
	txreq -hdr "be: three"
	rxresp
	expect resp.status == 201
	expect resp.body ~ "^3"
} -start
client c1 -wait
client c2 -wait
client c3 -wait
```
By storing this test in `test1.vtc` and running `varnishtest test1.vtc`, you'll see how it performs. **It will take more than 15 seconds** because the `set beresp.ttl = 0s;` causes so-called *serialization* on the waitinglist.

The servers contain artificial delay which will help us prove our point.

Here's the version with `set beresp.ttl = 120s;` :

```
varnishtest "Testing beresp.ttl=0s vs beresp.ttl=120s"
server s1 {
	rxreq
	delay 5
	txresp -status 201 -body "1\n"
} -start
server s2 {
	rxreq
	delay 5
	txresp -status 201 -body "2\n"
} -start
server s3 {
	rxreq
	delay 5
	txresp -status 201 -body "3\n"
} -start
varnish v1 -vcl+backend {
	sub vcl_backend_fetch {
		if (bereq.http.be == "two") {
			set bereq.backend = s2;
		} else if (bereq.http.be == "three") {
			set bereq.backend = s3;
		}
	}
	sub vcl_backend_response {
		if (beresp.status != 200 && beresp.status != 404) {
			set beresp.ttl = 120s;
			set beresp.uncacheable = true;
			return (deliver);
		}
	}
} -start
client c1 {
	txreq
	rxresp
	expect resp.status == 201
	expect resp.body ~ "^1"
} -start
client c2 {
	txreq -hdr "be: two"
	rxresp
	expect resp.status == 201
	expect resp.body ~ "^2"
} -start
client c3 {
	txreq -hdr "be: three"
	rxresp
	expect resp.status == 201
	expect resp.body ~ "^3"
} -start
client c1 -wait
client c2 -wait
client c3 -wait
```
Store this code in `test2.vtc` and run `varnishtest test2.vtc` to see the performance improvement. **This test should take a bit over 10 seconds**, because 2 out of 3 requests will be handled in parallel.

In a production environment where a lot more concurrency takes place, the performance impact will be more significant than in these isolated test cases.

> FYI: the `vcl_backend_fetch` part was added to the VCL solely to facilitate these tests. The code is required to support parallel processing of requests to the backend server.

**Comment by sidolov** [comment]
Changing priority since PR brings performance improvement for installations using Varnish

**Review by sidolov — APPROVED** [review]

**Comment by ihor-sviziev** [comment]
Unfortunately, it got added again in other places in https://github.com/magento/magento2/commit/026e5b29a5edfd619bbdea62d636b3cab2ea03b4 :(

File: app/code/Magento/PageCache/etc/varnish4.vcl
Comment (peterjaap) [reply app/code/Magento/PageCache/etc/varnish4.vcl:169]: Shouldn't this be set in the varnish5 and varnish6 vcl too?

**Pull request merged**
